### PR TITLE
Feature/100 0 n columns

### LINKED ISSUE #100: Agent: Refactor the code, Measures requiring 0-n columns, not exactly 1
## Background
Check the internal Agent's code, especially the `Measure` trait - it has `controlCol` which is a String - but the MeasureDTO has a Sequence of these columns. DTO is correct - some measures can be using 0 columns and some can be using n columns - the goal of this ticket is to refactor the Agent's code and its internal model if needed; DTO probably won't need t be changed.

### PR DESCRIPTION
**Pull request by salamonpavel**
Closes #100 
Introduces support for 0-n columns (depends on individual measure).

### CODE REVIEW

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala [hunk 1]
```diff
@@ -17,182 +17,156 @@
...
-  private val valueColumnName: String = "value"
+object AtumMeasure {
 
-  val supportedMeasures: Seq[MeasureType] = Seq(
-    RecordCount,
-    DistinctRecordCount,
-    SumOfValuesOfColumn,
-    AbsSumOfValuesOfColumn,
-    SumOfHashesOfColumn
+  val supportedMeasureNames: Seq[String] = Seq(
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:44]: why not to keep that? So we would offer strings as well as objects. No criticism, just want to know your reasoning

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala [hunk 2]
```diff
@@ -17,182 +17,156 @@
...
-import za.co.absa.spark.commons.implicits.StructTypeImplicits.StructTypeEnhancements
 
 /**
  *  Type of different measures to be applied to the columns.
  */
-sealed trait Measure extends MeasurementProcessor with MeasureType {
-  val controlCol: String
+sealed trait Measure {
+  val measureName: String
+  def controlColumns: Seq[String]
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:31]: I would even dare to say, let's rename it to `measuredColumns`. @benedeki WDYT?

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala [hunk 3]
```diff
@@ -17,182 +17,156 @@
...
+
+  case class DistinctRecordCount private (measureName: String, controlCols: Seq[String]) extends AtumMeasure {
+    require(controlCols.nonEmpty, "At least one control column has to be defined.")
 
-  case class DistinctRecordCount private (
-    controlCol: String,
-    measureName: String,
-    resultValueType: ResultValueType.ResultValueType
-  ) extends Measure {
+    private val columnExpression = countDistinct(col(controlCols.head), controlCols.tail.map(col): _*)
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:72]: I'm still not sure if this is what we want. I thought that it's about 'distinct -> count' operation, performed on a single column. Do you remember what was used in the old Atum & what can our users require @benedeki ?

**Comment by lsulak** [comment]
> Code is based off of feature/99-custom-measure-2.

Why not to repoint the PR then? It would be easier to do the review

**Comment by salamonpavel** [comment]
> Why not to repoint the PR then? It would be easier to do the review

Good point, will change the target branch to be the feature/99-custom-measure-2.

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala (hunk 3 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:72]: With the suggested code it can be performed on single column or multiple columns. If this particular measure is expected to work differently please let me know. @benedeki

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala (hunk 1 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:44]: I don't mind having the previous implementation once we really use that seq of objects. But until then it complicates matters (compilation) when one is refactoring the code.

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala (hunk 2 above)
Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:31]: Ok we agreed that it's what we want, but not now and here, but in https://github.com/AbsaOSS/atum-service/issues/116

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measurement.scala [hunk 4]
```diff
@@ -74,13 +74,13 @@ object Measurement {
...
+          MeasurementProvided[T](CustomMeasure(measureName, Seq(controlCol)), resultValue, ResultValueType.Long)
         case _: Double =>
-          MeasurementProvided[T](CustomMeasure(measureName, controlCol), resultValue, ResultValueType.Double)
+          MeasurementProvided[T](CustomMeasure(measureName, Seq(controlCol)), resultValue, ResultValueType.Double)
         case _: BigDecimal =>
-          MeasurementProvided[T](CustomMeasure(measureName, controlCol), resultValue, ResultValueType.BigDecimal)
+          MeasurementProvided[T](CustomMeasure(measureName, Seq(controlCol)), resultValue, ResultValueType.BigDecimal)
         case _: String =>
-          MeasurementProvided[T](CustomMeasure(measureName, controlCol), resultValue, ResultValueType.String)
+          MeasurementProvided[T](CustomMeasure(measureName, Seq(controlCol)), resultValue,  ResultValueType.String)
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/model/Measurement.scala:83]:
```suggestion
          MeasurementProvided[T](CustomMeasure(measureName, Seq(controlCol)), resultValue, ResultValueType.String)
```

**Review by lsulak — COMMENTED** [review]
I finished the review, I like the changes. There is only one tiny minor thing I found, and 1 thing regarding distinct count functionality. Otherwise LGTM and I will be happy to approve once these are addressed.

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala (hunk 1 above)
Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:44]: Okay

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala (hunk 3 above)
Comment (benedeki) [reply agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:72]: I like the universality of supporting multi-column uniqueness 👍

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala [hunk 5]
```diff
@@ -50,122 +49,124 @@ object AtumMeasure {
...
-  object AbsSumOfValuesOfColumn  {
-    private [agent] val measureName: String = "absAggregatedTotal"
+  object AbsSumOfValuesOfColumn {
+    private[agent] val measureName: String = "absAggregatedTotal"
     def apply(controlCol: String): AbsSumOfValuesOfColumn = AbsSumOfValuesOfColumn(measureName, controlCol)
   }
 
   case class SumOfHashesOfColumn private (measureName: String, controlCol: String) extends AtumMeasure {
-
+    private val columnExpression: Column = sum(crc32(col(controlCol).cast("String")))
```
Comment (benedeki) [inline agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:123]: Could this function also support a list of columns, concatenated.
Extra bonus is `*` for all columns. 😉

**Review by benedeki — COMMENTED** [review]
**To consider:**
I like class hierarchies. What you think of having 3 abstract classes
* `MeasureWithoutColumn`
* `MeasureOneColumn`
* `MeasureMultipleColumns`

File: agent/src/test/scala/za/co/absa/atum/agent/model/MeasurementTest.scala [hunk 6]
```diff
@@ -28,15 +28,15 @@ class MeasurementTest extends AnyFlatSpec with Matchers with SparkTestBase { sel
 
   "MeasurementProvided" should "be able to be converted to MeasurementProvided object when the result is Double" in {
```
Comment (miroslavpojer) [inline agent/src/test/scala/za/co/absa/atum/agent/model/MeasurementTest.scala:29]: Test speak about Double, but works with Big Decimals

**Review by miroslavpojer — COMMENTED** [review]
- [x] pulled
- [x] built
- [ ] test run
- [ ] code coverage - required merge of conflicting files, there were changes in tests
  - in current test implementation I see not covered paths

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala [hunk 7]
```diff
@@ -50,122 +49,124 @@ object AtumMeasure {
...
+        if (result == null) 0 else result
+      case _: StringType =>
         val value =
-          if (collected == null) new java.math.BigDecimal(0)
-          else collected.asInstanceOf[java.math.BigDecimal]
+          if (result == null) new java.math.BigDecimal(0)
+          else result.asInstanceOf[java.math.BigDecimal]
         value.stripTrailingZeros // removes trailing zeros (2001.500000 -> 2001.5, but can introduce scientific notation (600.000 -> 6E+2)
           .toPlainString // converts to normal string (6E+2 -> "600")
       case _ =>
```
Comment (benedeki) [inline agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:166]: For the record, this we might want to revisit and rethink. But not part of this PR. 😉

**Review by benedeki — COMMENTED** [review]
I think it's ok.
Just decide uf you want to add the hash function on multiple columns to add here or separate ticket+PR.

Oh and there's Miroslav's comment on one of the tests.

**Comment by salamonpavel** [comment]
I will re-point this PR to master instead of feature/99-custom-measure-2.

**Review by benedeki — APPROVED** [review]
